### PR TITLE
Optional IR integration into `sway-core`.

### LINKED ISSUE #211: Introduce intermediate representation to the compiler pipeline.
This is a biggish change, but is probably worth doing at this relatively early stage of the compiler's development.

Currently the pipeline is something like `parsing` -> `semantic analysis` -> `codegen`.  It pretty much goes from `source` -> `ast` -> `typed ast` -> `asm`.  The parsing and analysis could be called the front-end and the codegen could be called the back-end.

Most compilers these days have a 'middle-end' which is where the output from the front-end is an intermediate representation (IR) and that is in turn compiled into asm in the back-end.  It's particularly useful for compilers with several different language front-ends (C, C++, ObjC, ...) and several target back-end architectures (x86, ARM, MIPS, ...) allowing them to be bridged with a neutral common interface.

But for Sway/FuelVM we don't have multiple languages or targeted architectures, so do we really need a middle-end?

An advantage of using IR within the pipeline is to design it to be very easy to reason about, e.g., control flow analysis and data flow analysis, and to design it to be very easily manipulated while maintaining its integrity.  This in turn allows program transformations, especially optimisations to be performed more easily than might be possible otherwise.

The IR which many compiler toolkits use, in particular LLVM, is [static single assignment](https://en.wikipedia.org/wiki/Static_single_assignment_form) form.  The Wikipedia page there links to a bunch of benefits, mostly to do with optimisations.

Another basic benefit is the simplification of both the front-end and back-end.  The front-end needs to perform semantic analysis, but then to only produce correct IR, not necessarily efficient or even sensible IR.  And the back-end can translate almost directly from the SSA to ASM without needing to worry about optimisations.   Recently there have been a few bugs fixed related to the ASM gen in Sway which almost definitely would not have been a problem with the structure and rigor of an SSA IR to guide codegen.

My experience with LLVM also demonstrated to me the power of testing with SSA IR.  If we write a serialiser and deserialiser for the IR then testing can be more modular rather than end to end.   The output of the front-end only needs to be correct IR.  Our tests need not be clouded by what happens in the rest of the compiler.  And testing individual transformations/optimisations performed on the IR can be a matter of reading some  'poor' IR into the test and writing the 'improved' IR out, and confirming the transformation was correct, focusing only on the minimal case. 

It's less focused than unit testing but more focused than E2E testing. 

The Wikipedia page has some SSA links to read, and the great [C9X bib page](https://c9x.me/compile/bib/) has some links, in particular to the SSA book, which will be required reading prior to implementing this, I assume.  I need to read it too. :slightly_smiling_face:

### PR DESCRIPTION
**Pull request by otrho**
OK, this is the rest of the IR changes.  It's still a WIP to a degree, I'll list why below, but it's close enough that I'd like to merge it in.

Each commit is fairly specialised, it might be easier to review them individually.

Most of the work is in `sway-core/src/optimize.rs` and in `sway-core/src/asm_generation/from_ir.rs`.  There are some changes to `lib.rs` and to `forc` to allow a new `--use-ir` flag, for optionally turning compilaion via IR on.  It's obviously false/off by default.

I've changed a couple of ASM gen things too, just to make re-using them easier, same with a couple of `error.rs` changes.

So, while it should pass the E2E tests (last time I checked, which was actually a day or two ago, so that might not be quite true still...) it still needs a bit of work to be usable:
- _Spans_.  Because I started this before we had the new `Span`s I just did all the work without them.  Ideally each instruction in the IR would have a span, partly for better errors but mostly for source mapping.
- _Errors_.  Most of the new code just uses a `Result<_, String>`.  This should use a proper error struct rather than `String`.
- _Hairier crypto stuff_.  Contract calls are working, but I think they're a little slapdash still.  Predicates and libraries are pretty much untested and there might be other non-script cases which don't work.

I'm PRing this now before I go away partly so I don't have to come back and catch up to `master` and re-implement everything codegen-wise then.  If this is merged it can pretty much be ignored, as long as it compiles, and to make it compile it's good enough to just put in `todo!()` stubs if I'm needed to make sense of anything.

Closes #211.

### CODE REVIEW

File: forc/src/cli/commands/build.rs [hunk 1]
```diff
@@ -7,12 +7,18 @@ pub struct Command {
...
+    pub use_ir: bool,
     /// Whether to compile to bytecode (false) or to print out the generated ASM (true).
     #[structopt(long)]
     pub print_finalized_asm: bool,
     /// Whether to compile to bytecode (false) or to print out the generated ASM (true).
     #[structopt(long)]
     pub print_intermediate_asm: bool,
+    /// Whether to compile to bytecode (false) or to print out the generated IR (true).
+    #[structopt(long)]
+    pub print_ir: bool,
```
Comment (adlerjohn) [inline forc/src/cli/commands/build.rs:21]: Comment seems wrong. Aside: these flags that are dependent on each other remind me of https://www.youtube.com/watch?v=bnnacleqg6k.

File: forc/src/cli/commands/deploy.rs [hunk 2]
```diff
@@ -8,12 +8,18 @@ pub struct Command {
...
+    pub use_ir: bool,
     /// Whether to compile to bytecode (false) or to print out the generated ASM (true).
     #[structopt(long)]
     pub print_finalized_asm: bool,
     /// Whether to compile to bytecode (false) or to print out the generated ASM (true).
     #[structopt(long)]
     pub print_intermediate_asm: bool,
+    /// Whether to compile to bytecode (false) or to print out the IR (true).
+    #[structopt(long)]
+    pub print_ir: bool,
```
Comment (adlerjohn) [inline forc/src/cli/commands/deploy.rs:22]: Is this really needed for deploy? If they want to see the IR they can just run build.

File: forc/src/cli/commands/run.rs [hunk 3]
```diff
@@ -34,6 +38,10 @@ pub struct Command {
     #[structopt(long)]
     pub print_intermediate_asm: bool,
 
+    /// Whether to compile to bytecode (false) or to print out the IR (true).
+    #[structopt(long)]
+    pub print_ir: bool,
+
```
Comment (adlerjohn) [inline forc/src/cli/commands/run.rs:44]: Ditto.

File: forc/src/cli/commands/deploy.rs (hunk 2 above)
Comment (otrho) [reply forc/src/cli/commands/deploy.rs:22]: The `print_ir` flags are symmetrical to the already existing `print_intermediate_asm` and `print_finalized_asm` flags.

If they're set then the compilation will stop at the end and print out the ASM or IR.  They're useful for anyone, not just us.

There's no way to inspect the IR by just running build.  The `--use-ir` flag turns the IR compilation pipeline on, the `--print-ir` flag implies it.  I didn't bother documenting this as the `--use-ir` flag is temporary, it'll go away when we switch to IR full time.

Comment (sezna) [reply forc/src/cli/commands/deploy.rs:22]: That makes sense.

**Review by sezna — APPROVED** [review]
LFG. would you mind making an issue about the string return types and addressing it in short order? That kind of thing gets more painful to fix the longer we put it off

**Review by emilyaherbert — APPROVED** [review]

**Comment by emilyaherbert** [comment]
Looking awesome @otrho !

**Pull request merged**